packages/nx/src/project-graph/project-graph-builder.ts

```diff
@@ -72,29 +71,61 @@ export class ProjectGraphBuilder {
   }
 
   /**
-   * Adds a dependency from source project to target project
+   * Adds static dependency from source project to target project
    */
-  addImplicitDependency(
+  addStaticDependency(
     sourceProjectName: string,
-    targetProjectName: string
+    targetProjectName: string,
+    sourceProjectFile?: string
   ): void {
-    if (sourceProjectName === targetProjectName) {
-      return;
+    if (this.graph.nodes[sourceProjectName] && !sourceProjectFile) {
+      throw new Error(`Source project file is required`);
     }
-    if (!this.graph.nodes[sourceProjectName]) {
-      throw new Error(`Source project does not exist: ${sourceProjectName}`);
+    this.addDependency(
+      sourceProjectName,
+      targetProjectName,
+      DependencyType.static,
+      sourceProjectFile
+    );
+  }
+
+  /**
+   * Adds dynamic dependency from source project to target project
+   */
+  addDynamicDependency(
+    sourceProjectName: string,
+    targetProjectName: string,
+    sourceProjectFile: string
```

Comment:
If the above is optional, this should also be optional

Reply:
Not necessarily. Static dependency can exist between two external nodes where we have no source file information.

Dynamic dependency can only exist on project node detected by dynamic import on file. So there is always a file.

Reply:
@meeroslav I believe that this change made `package.json` dependencies dynamic (which breaks `nrwl/nx/enforce-module-boundaries`).

I think this is because [this code in `processPackageJson`](https://github.com/meeroslav/nx/blob/a922e64b4641e605d9d73870c76a89e93b60cf04/packages/nx/src/project-graph/build-dependencies/explicit-package-json-dependencies.ts#L84-L99) never creates a `static` type under any circumstances, and the above code always creates a dynamic dependency in the absence of a static one.

I'm not 100% sure that I'm right, but I was digging into lint failures caused by dynamic `package.json` file nodes in the dep graph, and it seems plausible that this change is the root cause.

Reply:
I am experiencing the same issue as @meeroslav after upgrading to version `15.8.1`

Reply:
This should be the fix https://github.com/nrwl/nx/pull/15381